### PR TITLE
8155902: DataOuputStream should clarify that it might write primitive types as multiple byte groups

### PR DESCRIPTION
**Pull request by bplb**
Add a disclaimer to `java.io.DataOutputStream` to the effect that it makes no guarantee as to how the underlying output stream actually writes the bytes provided to it.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8311882](https://bugs.openjdk.org/browse/JDK-8311882) to be approved

### Issues
 * [JDK-8155902](https://bugs.openjdk.org/browse/JDK-8155902): DataOuputStream should clarify that it might write primitive types as multiple byte groups (**Bug** - P4)
 * [JDK-8311882](https://bugs.openjdk.org/browse/JDK-8311882): DataOuputStream should clarify that it might write primitive types as multiple byte groups (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14791/head:pull/14791` \
`$ git checkout pull/14791`

Update a local copy of the PR: \
`$ git checkout pull/14791` \
`$ git pull https://git.openjdk.org/jdk.git pull/14791/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14791`

View PR using the GUI difftool: \
`$ git pr show -t 14791`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14791.diff">https://git.openjdk.org/jdk/pull/14791.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14791#issuecomment-1624351015)

### CODE REVIEW

**Comment by bplb** [comment]
It might be equally reasonable simply to resolve the issue as "Not an Issue." A CSR will be created if that is not the case.

**Comment by djelinski** [comment]
This issue is reported against JDK 8 and 9. It was fixed in 16 with 17f04fc9. I think you can close it as duplicate.

File: src/java.base/share/classes/java/io/DataOutputStream.java [hunk 1]
```diff
@@ -37,6 +37,11 @@
  * thread then access to the data output stream should be controlled
  * by appropriate synchronization.
  *
+ * @apiNote
+ * The write methods of this class merely delegate to the methods of
+ * the underlying output stream, hence there is no guarantee as to
+ * the details of how the bytes are written.
+ *
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/io/DataOutputStream.java:44]: I'd drop "merely".

I would not want to give any assurances about whether the bytes of an encoded primitive were written individually or as a group.  
For the primitive types the value is encoded and written as one or more bytes to the underlying stream.

**Comment by bplb** [comment]
> This issue is reported against JDK 8 and 9. It was fixed in 16 with [17f04fc](https://github.com/openjdk/jdk/commit/17f04fc9e755a83727ac9d13a39b487f99311efd). I think you can close it as duplicate.

Thanks. I saw that change go by but did not connect the dots. Was it actually verified that things are not split up into two or more packets? It seems that `write(2)` could in fact return a value less than the number of bytes requested.

File: src/java.base/share/classes/java/io/DataOutputStream.java (hunk 1 above)
Comment (bplb) [reply src/java.base/share/classes/java/io/DataOutputStream.java]: > I would not want to give any assurances about whether the bytes of an encoded primitive were written individually or as a group.

Would it be helpful to add further verbiage?
```
For example, the sequence of bytes comprising the primitive data type may be written as more than one packet of bytes.
```

**Comment by djelinski** [comment]
> Was it actually verified that things are not split up into two or more packets? It seems that `write(2)` could in fact return a value less than the number of bytes requested.

Yes, the test case included in JBS sends all 4 bytes in a single packet now.
We can't guarantee that every writeInt will always send all bytes in the same packet, but at least we won't send single-byte packets.

File: src/java.base/share/classes/java/io/DataOutputStream.java (hunk 1 above)
Comment (RogerRiggs) [reply src/java.base/share/classes/java/io/DataOutputStream.java]: "packet" is network centric terminology.
Perhaps something like:  "may be written individually or in any grouping of bytes"

Comment (bplb) [reply src/java.base/share/classes/java/io/DataOutputStream.java]: Commit 43c392e8c797f5a5c29a67e87f350e8dc9c5b190 changed it to
```
The write methods of this class delegate to the methods
of the underlying output stream, hence there is no 
guarantee as to the details of how the bytes are written.
```

**Comment by AlanBateman** [comment]
In passing, we should probably look into the inconsistency in the synchronization. The write methods are synchronized, the writeXXX methods and other methods are not. The class specified that it is not safe for use by concurrent threads without appropriate synchronization.

File: src/java.base/share/classes/java/io/DataOutputStream.java [hunk 2]
```diff
@@ -37,6 +37,11 @@
  * thread then access to the data output stream should be controlled
  * by appropriate synchronization.
  *
+ * @apiNote
+ * The write methods of this class delegate to the methods of
+ * the underlying output stream, hence there is no guarantee as to
+ * the details of how the bytes are written.
+ *
```
[the remark this answers is not in the document]
Comment (bplb) [reply src/java.base/share/classes/java/io/DataOutputStream.java]: Changed accordingly in f89aa63ac2b03e66ae3f7238e4decbb0d28794de.

File: src/java.base/share/classes/java/io/DataOutputStream.java [hunk 3]
```diff
@@ -30,7 +30,11 @@
 /**
  * A data output stream lets an application write primitive Java data
  * types to an output stream in a portable way. An application can
- * then use a data input stream to read the data back in.
+ * then use a data input stream to read the data back in. A data output
+ * stream wraps another output stream and delegates writing bytes to the
+ * write methods of that output stream. Hence writing data consisting of
+ * more than a single byte may cause several writes to the underlying
+ * output stream.
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/DataOutputStream.java:37]: This looks much better than the apiNote. One suggestion is to drop "Hence", I don't think it is needed here.

**Review by djelinski — APPROVED** [review]
LGTM.

**Review by RogerRiggs — APPROVED** [review]

**Pull request closed without merging**